### PR TITLE
add status of competition in model, set status active for new competi…

### PR DESCRIPTION
**Pull request by GuskaAW**
Add in competition model:
- Enum for competition status (active, future, done)
- Add set_status_to_new_competition (active)
- Add data_migration for competitions with status nil => Change status to active. 
- Add CheckCompetitionStatusService to check if every task is done. If so, change competitions status to done. 
- Set permissions in controllers and views for competitions with different statuses. 
 - For competition status active o future:
      competition: destroy, edit, update
      tasks: edit, update, change_task_name_or_points, delete
 - For competition status active:
      tasks:  change_status_to_done, add_points_to_user, take_task, take_task, give_back_task 
- Add Congratulation email when compatition has status done
- Add service tasks_to_string_service for 2 mailers (1. competition done - run from controller and 2. weekly_competition_summary run from service)
- Add possibility to set competition status manually from competition-edit on select from enum
After review:
- Add in competition Model 2 methods status_done? and status_active?
- Add in task controller private methods competition_status_done? and _active? and changes with it
- Add in competition controller private method competition_status_done? and use it, changes with it 
- tasks and competition views changed with this methods
- Add competition_done_service for congratulation email if competition done.

### CODE REVIEW

File: app/views/competitions/index.html.erb [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+  </thead>
   
   <% @competitions.each do |competition| %>
   <tr>
       <td><%= competition.id %><td>
       <td><%= competition.created_at %></td>
+      <td><% if competition.status %>
+          <%= competition.status %>
+      <% else %>
+        <%= "without status" %>
```
Comment (yzmopolis) [inline app/views/competitions/index.html.erb:20]: We shouldn't have a situation `without status` if you have a data migration.

File: config/initializers/sidekiq.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 schedule_file = "config/schedule.yml"
 if File.exist?(schedule_file) && Sidekiq.server?
-	#Sidekiq::Cron::Job.create(name: 'WeeklyService - every 1min', cron: '*/1 * * * *', class: 'WeeklyService') # execute at every 5 minutes, ex: 12:05, 12:10, 12:15...etc
-    Sidekiq::Cron::Job.load_from_hash YAML.load_file(schedule_file)
+	Sidekiq::Cron::Job.create(name: 'WeeklyService - every 1min', cron: '*/1 * * * *', class: 'WeeklyService') # execute at every 5 minutes, ex: 12:05, 12:10, 12:15...etc
+    #Sidekiq::Cron::Job.load_from_hash YAML.load_file(schedule_file)
     #Sidekiq::Cron::Job.destroy_all! #for deleting all jobs
```
Comment (yzmopolis) [inline config/initializers/sidekiq.rb:5]: Great job with that 🎉

File: db/data/20200520130326_change_nil_in_competition_status_to_active.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class ChangeNilInCompetitionStatusToActive < ActiveRecord::Migration[5.2]
+  def up
+  	Competition.where(status: nil).update_all(status: "active")
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
+  end
+end
```
Comment (yzmopolis) [inline db/data/20200520130326_change_nil_in_competition_status_to_active.rb:9]: Great job!

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -39,72 +41,96 @@ def create
...
-        redirect_to competition_path(@competition), notice: "Task '#{@task.name}' updated!"
+      a = change_task_name_or_points(task_params)
+
+     if current_ability.can?(:edit, @competition) 
+        if @task.update!(task_params)
+          if Rails.env.development?
+            if a == "task name was changed"
+              ChangeTaskNameOrPointsMailer.notify_user_name(@user, @player_1, @player_2, @task, @competition).deliver
+            else a == "task points were changed "
+              ChangeTaskNameOrPointsMailer.notify_user_points(@user, @player_1, @player_2, @task, @competition).deliver
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/tasks_controller.rb:55]: But the condition is not the same (points or name is changed)

File: app/services/check_competition_status_service.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+       if number_of_tasks == number_of_tasks_done
+         competition.update!(status: "done")
+       else
+         competition.update!(status: "active")
+       end
+      end
+#or less code
+# competition.tasks.pluck(:status).all? {|status| status == "done" } ? 
+# competition.update!(status: "done") : competition.update!(status: "active")
+
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/services/check_competition_status_service.rb:15]: Thx. Done

File: app/controllers/competitions_controller.rb [hunk 6]
```diff
@@ -57,45 +59,49 @@ def create
...
+				task_schimmels = TaskSchimmel.where(id: task_schimmels_params)
+				@tasks = TaskCreatorService.new.call(competition_id: @competition.id, task_schimmels: task_schimmels)
+				if Rails.env.development?
+					if a == "updated competition duration and new tasks"
+						ChangesCompetitionDurationOrTasksMailer.notify_user_duration_and_task(@user, @player_1, @player_2, @tasks, @duration, @competition).deliver	
+					elsif a == "updated new tasks"
+		    			ChangesCompetitionDurationOrTasksMailer.notify_user_new_task(@user, @player_1, @player_2, @tasks, @competition).deliver
+					else a == "updated competition duration"
+						ChangesCompetitionDurationOrTasksMailer.notify_user_duration(@user, @player_1, @player_2, @duration, @competition).deliver		
+					end
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/competitions_controller.rb:78]: I did it with "a" variable to make it more readable. The code is a log one and with an "a" I find it quickly. When I name it "change" its to much to read for me :)

File: app/controllers/competitions_controller.rb [hunk 7]
```diff
@@ -21,12 +21,14 @@ def new
 	end
 
 	def edit
+		if @competition.status != "done"
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/competitions_controller.rb]: Done, with new methods in competition model

File: app/controllers/tasks_controller.rb [hunk 8]
```diff
@@ -39,72 +41,96 @@ def create
...
-      redirect_to tasks_path(@task), notice: "You are not a player in competition #{@task.competition_id}. You can't make changes"
     end
   end
 
   def change_task_name_or_points(task_params)
-    if task_params[:name]
-      return "task name was changed" 
-    else task_params[:points]
-      return "task points were changed "
+    if @competition.status != "done"
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/tasks_controller.rb]: done

File: app/controllers/tasks_controller.rb [hunk 9]
```diff
@@ -39,72 +41,96 @@ def create
...
+    @competition_status = CheckCompetitionStatusService.new.check_competition_status(competition: @competition)
     redirect_to users_competition_path(@competition), notice: "Task '#{@task.name}' updated to taken!"
   end
-
+  
   def change_status_done
-    add_points_to_user(@task.points) if @task.status == "taken"
-    @task.update!(status: "done")
-    redirect_to users_competition_path(@competition), notice: "Task '#{@task.name}' updated! to tasks done"
+    if @competition.status == "active"
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/tasks_controller.rb]: done, with model method and private method

File: app/controllers/tasks_controller.rb [hunk 10]
```diff
@@ -39,72 +41,96 @@ def create
...
-    @player_1 = @competition.player_1
-    @player_2 = @competition.player_2
-    current_user.points += points
-    current_user.save!
-    @user = current_user
-    
-    if Rails.env.development?
-      #GetPointsEmailMailer.notify_user(@user, @player_1, @player_2, @task).deliver  
-      GetPointsEmailWorker.perform_in(2.second, @user.name, @user.points, @player_1.email, @player_2.email, @task.name, @task.points)    
+    if @competition.status == "active"
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/tasks_controller.rb]: done

File: app/controllers/tasks_controller.rb [hunk 11]
```diff
@@ -39,72 +41,96 @@ def create
...
+      if @competition.status == "done"
+        player_1_tasks_done = @competition.player_1.tasks.done.where(competition_id: @competition.id)
+        player_1_tasks_done_string = TasksToStringService.new.tasks_to_string(player_1_tasks_done)
+        player_1_tasks_done_sum = @competition.player_1.tasks.done.where(competition_id: @competition.id).count
+
+        player_2_tasks_done = @competition.player_2.tasks.done.where(competition_id: @competition.id)
+        player_2_tasks_done_string = TasksToStringService.new.tasks_to_string(player_2_tasks_done)
+        player_2_tasks_done_sum = @competition.player_2.tasks.done.where(competition_id: @competition.id).count
+    
+        CongratulationCompetitionDoneEmailWorker.perform_in(2.second,@player_1.name, @player_1.email, @player_1.points, player_1_tasks_done_string, player_1_tasks_done_sum, @player_2.name, @player_2.email, @player_2.points, player_2_tasks_done_string, player_2_tasks_done_sum, @competition.id, @duration)
```
[the remark this answers is not in the document]
Comment (GuskaAW) [reply app/controllers/tasks_controller.rb]: Thank You! Its a very good Idea. Done.

**Pull request merged**